osu.Game/GameModes/Play/PlaySongSelect.cs

```diff
@@ -4,37 +4,190 @@
 using System;
 using System.Collections.Generic;
 using osu.Framework.Configuration;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Beatmaps;
+using osu.Game.Beatmaps.IO;
 using osu.Game.GameModes.Backgrounds;
 using osu.Framework;
+using osu.Game.Database;
+using osu.Framework.Graphics.Primitives;
+using System.Linq;
+using OpenTK;
+using OpenTK.Graphics;
+using osu.Framework.Graphics.Textures;
+using osu.Framework.Graphics.UserInterface;
+using System.Threading.Tasks;
+using System.Diagnostics;
 
 namespace osu.Game.GameModes.Play
 {
-    class PlaySongSelect : GameModeWhiteBox
+    public class PlaySongSelect : OsuGameMode
     {
         private Bindable<PlayMode> playMode;
-
+        private BeatmapDatabase beatmaps;
+        private BeatmapSetInfo selectedBeatmapSet;
+        private BeatmapInfo selectedBeatmap;
+        // TODO: use currently selected track as bg
         protected override BackgroundMode CreateBackground() => new BackgroundModeCustom(@"Backgrounds/bg4");
+        private ScrollContainer scrollContainer;
+        private FlowContainer setList;
 
-        protected override IEnumerable<Type> PossibleChildren => new[] {
-                typeof(ModSelect),
-                typeof(Player)
-        };
+        private void selectBeatmapSet(BeatmapSetInfo beatmapSet)
+        {
+            selectedBeatmapSet = beatmapSet;
+            foreach (var child in setList.Children)
+            {
+                var childGroup = child as BeatmapGroup;
+                if (childGroup.BeatmapSet == beatmapSet)
+                {
+                    childGroup.Collapsed = false;
+                    selectedBeatmap = childGroup.SelectedBeatmap;
+                }
+                else
+                    childGroup.Collapsed = true;
+            }
+        }
+        
+        private void selectBeatmap(BeatmapSetInfo set, BeatmapInfo beatmap)
+        {
+            selectBeatmapSet(set);
+            selectedBeatmap = beatmap;
+        }
+
+        private Stopwatch watch = new Stopwatch();
+
+        private void addBeatmapSet(BeatmapSetInfo beatmapSet)
+        {
+            watch.Reset();
+            watch.Start();
+            beatmapSet = beatmaps.GetWithChildren<BeatmapSetInfo>(beatmapSet.BeatmapSetID);
+            beatmapSet.Beatmaps.ForEach(b => beatmaps.GetChildren(b));
+            beatmapSet.Beatmaps = beatmapSet.Beatmaps.OrderBy(b => b.BaseDifficulty.OverallDifficulty)
+                .ToList();
+            Scheduler.Add(() =>
+            {
+                var group = new BeatmapGroup(beatmapSet);
+                group.SetSelected += selectBeatmapSet;

```

Comment:
i'd consider making these non-events and using object initialiser to bind them (just for code elegance). pretty sure we'll only be binding from one place

```
var group = new BeatmapGroup(beatmapSet)
                {
                    SetSelected = selectBeatmapSet,
                    BeatmapSelected = selectBeatmap
                };
```